server/app/survey.go

```diff
@@ -50,3 +64,211 @@ func (a *UserSurveyApp) StopSurvey(surveyID string) error {
+		return errors.Wrap(appErr, "SendSurvey: failed to get user from ID: "+userID)
+	}
+
+	// open a DM between the bot and the user
+	botUserDM, appErr := a.api.GetDirectChannel(user.Id, a.botID)
+	if appErr != nil {
+		errMsg := fmt.Sprintf("SendSurvey: failed to create DM between survey bot and user, botID: %s, userID: %s, error: %s", a.botID, userID, appErr.Error())
+		a.api.LogError(errMsg)
+		return errors.Wrap(errors.New(appErr.Error()), errMsg)
+	}
+
+	postMessage := fmt.Sprintf(":wave: Hey @%s! %s", user.Username, survey.SurveyQuestions.SurveyMessageText)
+	post := &mmModal.Post{
+		UserId:    a.botID,
+		Message:   postMessage,
+		ChannelId: botUserDM.Id,
+		Type:      surveyPostType,
+	}
+
+	questionsJSON, err := json.Marshal(survey.SurveyQuestions)
+	if err != nil {
+		a.api.LogError("SendSurvey: failed to marshal survey questions for inserting into post", "error", err.Error())
+		return errors.Wrap(err, "SendSurvey: failed to marshal survey questions for inserting into post")
+	}
+
+	post.AddProp("survey_questions", string(questionsJSON))
+
+	_, appErr = a.api.CreatePost(post)
+	if appErr != nil {
+		a.api.LogError("SendSurvey: failed to create survey post for user", "userID", userID, "error", appErr.Error())
+		return errors.Wrap(appErr, "SendSurvey: failed to create survey post for user")
+	}
+
+	if err := a.setSurveySentToUser(userID, survey.ID); err != nil {
+		return errors.Wrap(err, "SendSurvey: failed to mark survey set to user")
+	}
+
+	return nil
+}
+
+func (a *UserSurveyApp) ensureSurveyBot() error {
+	if a.botID != "" {
+		return nil
+	}
+
+	bot := &mmModal.Bot{
+		Username:    "feedbackbot",
+		DisplayName: "Feedbackbot",
+		Description: "Created by User Survey plugin",
+	}
+
+	botID, err := a.apiClient.Bot.EnsureBot(bot, pluginapi.ProfileImageBytes(assets.BotIcon))
+	if err != nil {
+		a.api.LogError("failed to ensure feedback bot", "error", err.Error())
+		return errors.Wrap(err, "ensureSurveyBot: failed to ensure feedback bot")
+	}
+
+	a.botID = botID
+	return nil
+}
+
+func (a *UserSurveyApp) userInExcludedTeams(userID string, survey *model.Survey) (bool, error) {
+	// check in cache first
+	result, ok, err := a.getCachedUserNotInFilteredTeams(userID, survey.ID)
+	if err != nil {
+		return false, err
+	}
+
+	if ok {
+		// if there was a cache hit, just return the cached result
+		return result, nil
+	}
+
+	// compute data on cache miss
+	teams, appErr := a.api.GetTeamsForUser(userID)
+	if appErr != nil {
+		a.api.LogError("userInExcludedTeams: failed to get user teams", "userID", userID, "error", appErr.Error())
+		return false, errors.Wrap(errors.New(appErr.Error()), "userInExcludedTeams: failed to get user teams")
+	}
+
+	filteredTeamMap := map[string]bool{}
+	for _, teamID := range survey.ExcludedTeamIDs {
+		filteredTeamMap[teamID] = true
+	}
+
+	userMemberOfFilteredTeam := false
+
+	for _, team := range teams {
+		if _, ok := filteredTeamMap[team.Id]; ok {
+			userMemberOfFilteredTeam = true
+			break
+		}
+	}
```

Comment:
@harshilsharma63 this seems to imply that if a user is a member of teams: Team A, Team B, Team C and only Team C is on the excluded teams, they will not receive the survey despite being in teams A and B that are not excluded. Is that the desired state?

Reply:
@esarafianou your'e absolutely right, they won't receive the survey. This is indeed the desired behaviour.

This was discussed here as well for more context - https://hub.mattermost.com/private-core/pl/ncrx1rchfjbbjb8hrdatc3qony

Reply:
Sounds good, thanks for the context!